### DISCUSSION TITLE
variadic / array with MapRequestPayload

### DISCUSSION THREAD

**Discussion by themark147** [symfony/symfony#50661] [Q&A] [answered]
hey

I am still trying to figure out how to use MapRequestPayload with variadic / array of objects

eg. my payload
```
[
  {"number": 1},
  {"number": 2}
]
```
```
final class ObjectDto
{
    public function __construct(
        public readonly int $number
    )
    {
    }
}
```

controller 

```
__invoke(#[MapRequestPayload] ...ObjectDto $objects)
```

or its not implemented yet? is there any possibility to do it custom way?

thanks :hugs:

**Comment by ghost** [comment]
Have you tried it works? If it throws exception because of variadic parameter then it's not implemented. MapRequestPayload just deserializes request content and that's all. So if your serializer is able to deserialize this then it should work.

**Reply by ghost** [reply]
On second thoughts I can see this will not work without additional work on Symfony side. The `$type` argument passed to `Serializer::deserialize()` https://github.com/symfony/symfony/blob/85182e7e9d71a3246f8447ef8f55c1975545fe75/src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/RequestPayloadValueResolver.php#L182
is taken from `ArgumentMetadata` but this call does not consider `ArgumentMetadata::$isVariadic`.

I think this could be solved by changing this line to
```php
if ($argumentMetadata->isVariadic()) {
  return $this->serializer->deserialize($data, 'array<'.$type.'>', $format, self::CONTEXT_DESERIALIZE + $attribute->serializationContext);
}
```

**Reply by stof** [reply, marked as answer]
For variadic arguments, the resolve method would have to return an array with multiple arguments in it. As it delays the parsing until later (for a reason that is not clear to me), it cannot perform decoding to multiple objects.

I think the `resolve` method should throw an exception if used on a variadic argument with the current implementation.

**Reply by themark147** [reply]
thanks for your answers ... so only option is use that array in some DTO and that DTO use in controller param right?

can somebody help me with creating issue or PR to throw exception on array / variadic argument?

thanks :hugs:

**Reply by ghost** [reply]
Creating issue is very easy. Just describe the problem and wait for feedback. To create a PR you have to fork the repository, introduce changes to code, create a PR from your forked repository to symfony/symfony repository and wait for the code review. Github has it well documented how to contribute to the public repositories.

**Reply by themark147** [reply]
ok than thanks I will try

**Reply by Asymptota** [reply]
> ok than thanks I will try

Hello @themark147

Have you created an issue or PR related to this topic?

I have found myself in same situation and looking for solid solution.

**Reply by themark147** [reply]
Hi @Asymptota, variadic params are not supported ... PR: https://github.com/symfony/symfony/pull/50673/files + issue: https://github.com/symfony/symfony/issues/50668 ... have to use array of objects instead of variadic approach

**Comment by DmitriyDodon** [comment]
Adding doc /** @var ObjectDto[] */ helped in my case.

**Comment by zangetsu02** [comment]
Hello,
I am having the same issue,
```json
[
  {"number": 1},
  {"number": 2}
]
```


```php
	
class Dto {
	public function __constructor(
		public int $number;
	) {}
}

// Controller

public function index(#[MapRequestPayload] Dto $dto) {...}
```

is there a way of using the MapRequestPayload like this?

**Reply by themark147** [reply]
u cannot pass payload as array cause controller expects: {"number": 1}
